src/main/java/uk/gov/ons/ctp/response/action/service/ActionService.java

```diff
@@ -173,25 +200,129 @@ public Action createAction(final Action action) {
-        actionPlanRepository.findByActionPlanPK(actionPlanJob.getActionPlanFK());
+  private boolean hasRuleTriggered(ActionRule rule) {
     final Timestamp currentTime = new Timestamp((new Date()).getTime());
+    Timestamp triggerDateTime =
+        Timestamp.valueOf(
+            LocalDateTime.ofInstant(rule.getTriggerDateTime().toInstant(), ZoneOffset.UTC));
+    return triggerDateTime.before(currentTime);
+  }
+
+  private void createActions(ActionCase actionCase, ActionRule actionRule) {
+    ActionType actionType = actionTypeRepo.findByActionTypePK(actionRule.getActionTypeFK());
+
+    // If creating actions for business respondents, create an action per respondent
+    // Else create single action
+    if (actionCase.getSampleUnitType().equals(SampleUnitDTO.SampleUnitType.B.toString())
+        && (actionType.getActionTypePK() == 5 || actionType.getActionTypePK() == 7)) {
 
-    actionRepo
-        .findPotentialActionsActiveDate(actionPlan.getActionPlanPK(), currentTime)
+      List<Association> enrolledAssociations = respondentsEnrolledOnCase(actionCase);
+      if (enrolledAssociations.isEmpty()) {
+        log.error(NO_RESPONDENTS_FOR_SURVEY);
+        throw new IllegalStateException(NO_RESPONDENTS_FOR_SURVEY);
+      }
+
+      enrolledAssociations.forEach(
+          association ->
+              createAction(
+                  actionCase, actionRule, actionType, UUID.fromString(association.getPartyId())));
+
+    } else {
+      createAction(actionCase, actionRule, actionType, actionCase.getPartyId());
+    }
+  }
+
+  private List<Association> respondentsEnrolledOnCase(ActionCase actionCase) {
+    PartyDTO businessParty =
+        partySvcClientService.getParty(actionCase.getSampleUnitType(), actionCase.getPartyId());
+    CollectionExerciseDTO collectionExercise =
+        collectionExerciseClientService.getCollectionExercise(actionCase.getCollectionExerciseId());
+    return associationsEnrolledForSurvey(businessParty, collectionExercise.getSurveyId());
+  }
+
+  private List<Association> associationsEnrolledForSurvey(PartyDTO party, String surveyId) {
+    return party
+        .getAssociations()
         .stream()
-        .map(pa -> Action.fromPotentialAction(pa, currentTime))
-        .forEach(a -> actionRepo.save(a));
-    actionRepo.flush();
+        .filter(association -> isAssociationEnabledForSurvey(association, surveyId))
+        .collect(Collectors.toList());
+  }
+
+  private boolean isAssociationEnabledForSurvey(Association association, String surveyId) {
+    return association
+        .getEnrolments()
+        .stream()
+        .anyMatch(
+            enrolment ->
+                enrolment.getSurveyId().equals(surveyId)
+                    && enrolment.getEnrolmentStatus().equalsIgnoreCase(ENABLED));
+  }
+
+  private void createAction(
+      ActionCase actionCase, ActionRule actionRule, ActionType actionType, UUID partyId) {
+    if (actionRepo.findOneByCaseIdAndActionRuleFKAndPartyId(
```

Comment:
if this is the only place the repo method is used, you should change it to `exists`